s3torchbenchmarking/src/s3torchbenchmarking/dcp/benchmark.py

```diff
@@ -0,0 +1,156 @@
+
+        # Dump the multiprocessing Queue's content into a list, and compute its "ptp" (peak to peak, i.e.,
+        # diff(max, min)) metric.
+        collector: List[float] = []
+        while not save_timestamps.empty():
+            collector.append(save_timestamps.get())
+        ptp_save_times.append(np.ptp(collector))
+
+    save_results(
+        cfg,
+        benchmark_model,
+        ptp_save_times=ptp_save_times,
+        processing_times=processing_times,
+        monitor=monitor,
+    )
+
+
+def validate_world_size(world_size: int) -> int:
+    """Enforce `world_size` to be within the current node's capacity.
+
+    FIXME: only works when the backend is "nccl"; what about "gloo" and CPUs?
+    """
+    device_count = torch.cuda.device_count()
+    if world_size > device_count:
+        logger.warning(
+            "Received a `world_size` of %i, while only %i are available: decreasing to %i",
+            world_size,
+            device_count,
+            device_count,
+        )
+        return device_count
+    return world_size
+
+
+def get_writer(cfg: DictConfig, suffix: str) -> FileSystemWriter:
+    """Instantiate a checkpoint writer based on the input config."""
+    if cfg.checkpoint.storage == "disk":
+        local_path = Path(cfg.path) / suffix
+        logger.info("Saving checkpoint to %s (local disk)...", local_path)
+        return dcp.FileSystemWriter(local_path, thread_count=cfg.thread_count)
+    elif cfg.checkpoint.storage == "s3":
+        uri = build_checkpoint_uri(cfg.s3.uri, suffix)
+        logger.info("Saving checkpoint to %s (S3)...", uri)
+        return S3StorageWriter(cfg.s3.region, uri, thread_count=cfg.thread_count)
+    raise ValueError(f"Storage writer {cfg.checkpoint.storage} not supported")
+
+
+def build_checkpoint_uri(s3_uri: str, suffix: str) -> str:
+    suffix = suffix.lstrip("/")
+    return s3_uri + suffix if s3_uri.endswith("/") else s3_uri + "/" + suffix
+
+
+def setup(backend: str, world_size: int, rank: int) -> None:
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+    dist.init_process_group(backend=backend, world_size=world_size, rank=rank)
+
+
+def run(
+    rank: int,  # needs to be passed first (provided by `multiprocessing.spawn` automatically)
+    backend: str,
+    epochs: int,
+    model: Module,
+    world_size: int,
+    storage_writer: FileSystemWriter,
+    save_timestamps: Queue,
+) -> None:
+    """Execute the actual code for checkpoint saving.
+
+    This function is meant to be executed in subprocesses."""
+
+    # FIXME: configure logging in subprocess accordingly
+
+    torch.cuda.set_device(rank)
+    setup(backend=backend, world_size=world_size, rank=rank)
+    rank = dist.get_rank()
+
+    device_id = rank % torch.cuda.device_count()
+    model.to(device_id)
+
+    ddp_model = DistributedDataParallel(model, device_ids=[device_id])
+
+    start_time = perf_counter()
+    for i in range(epochs):
```

Comment:
Adding a loop here does not make much sense to me, we should instead add the epochs in the place where the we are spawning as we are trying to find a mean run time over 5 epochs.